### ISSUE TITLE
misleading javadoc for Quaternion fromAngles()

### ISSUE THREAD

**Issue by stephengold** [jMonkeyEngine/jmonkeyengine#1388]
Currently the javadoc for Quaternion fromAngles() says:
```
Note that we are applying in order: (y, x, z) aka (yaw, pitch, roll)
but we've ordered them in x, y, and z for convenience.
```

1. I believe that the angles are actually applied in the order (x, z, y) not (y, x, z). One could write a short console app to test this.
2. There are actually six possible assignments of yaw/pitch/roll axes to x/y/z, so the information about yaw/pitch/roll is unhelpful.

**Comment by pspeed42** [comment]
It all depends on how you interpret the order.  If you tried to give instructions for someone to turn their head to represent the same orientation, you'd tell them to:
1. rotate your head left/right (yaw)
2. look up down (pitch)
3. then tilt your head (roll)

Also if you were to construct a scene graph of:
Node parent = new Node("parent");
Node child = new Node("child");
parent.attachChild(child);
Node grandChild = new Node("grandchild");
child.attachChild(grandchild);
parent.rotate(0, yaw, 0);
child.rotate(pitch, 0, 0);
grandchild.rotate(0, 0, roll);

Or for separate quaternions:
Quaternion yaw = new Quaternion().fromAngles(0, y, 0);
Quaternion pitch = new Quaternion().fromAngles(x, 0, 0);
Quaternion roll = new Quaternion().fromAngles(0, 0, z);
Quaternion q = yaw.mult(pitch).mult(roll);

So you can see why it's tempting to say the order is yaw, pitch, roll.... (It's also the order you'd move a joystick to simulate the same orientation.)

It's only from the child looking up to the parent that the order is reversed.  And whether that's the predominate way of looking at things tends to settle on whether preorder or postorder multiplication is used.  JME goes parent.mult(child) and so I think ordering makes more sense as yaw, pitch, roll.  parent.mult(child) is not universal, though... just the way JME does it.

**Comment by stephengold** [comment]
If the code is broken, it would be bad to change it at this point. We should document what it actually does, not what it "should" do.

**Comment by pspeed42** [comment]
Well, I tested it and it turns out that it is in fact broken.  Roll is essentially useless right now because it changes what pitch means but not what yaw means.  This matches the control scheme of nothing.  So it is yaw, roll, pitch at the moment (with top-down ordering).  Fortunately, the FPS style control schemes that would accumulate euler angles in this way are also not likely using roll.

Meaning the only folks using roll are probably ones who hand tweaked an orientation through a UI and fixing this would break them.

So we should fix the javadoc and point out that this is kind of broken and let people know if they want FPS-style ordering with roll then they have to implement it themselves.

I'd also argue for deprecating this method in favor of ones where order is specific for some common schemes.  I wonder if toAngles() is similarly incorrectly ordered or has the desired ordering... which may explain some of why folks have trouble plugging the values of one into the other beyond the normal euler ambiguity.

**Comment by tlf30** [comment]
I just ran into this. I was pulling my hair out trying to figure out why every load of my objects from my database caused them to rotate.

**Comment by pspeed42** [comment]
I think because it IS a breaking change, that another option would be to create methods like fromAnglesYXZ(), etc. and deprecate this one.

...otherwise, I generally agree that it's wrong.  It's written for an x=forward coordinate system where JME uses a z=forward coordinate system.

**Comment by tlf30** [comment]
So I wrote a test this morning while drinking my coffee. From zero, each axis does work correctly, but once changing multiple angles, it quickly becomes apparent that the order the rotation angles are applied is wrong.
```
package io.tlf.jme.test;

import com.jayfella.jme.jfx.JavaFxUI;
import com.jme3.app.SimpleApplication;
import com.jme3.input.KeyInput;
import com.jme3.input.controls.ActionListener;
import com.jme3.input.controls.KeyTrigger;
import com.jme3.math.ColorRGBA;
import com.jme3.math.Quaternion;
import com.jme3.math.Vector3f;
import com.jme3.scene.Node;
import com.jme3.scene.Spatial;
import javafx.beans.value.ChangeListener;
import javafx.beans.value.ObservableValue;
import javafx.event.ActionEvent;
import javafx.event.EventHandler;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.control.Slider;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;

public class RotationTest extends SimpleApplication {

    private boolean cursor = false;

    @Override
    public void simpleInitApp() {
        this.getViewPort().setBackgroundColor(ColorRGBA.White);
        //Import char
        Spatial player = assetManager.loadModel("Models/mixamo/strut_walking.j3o");

        //UI
        JavaFxUI.initialize(this);
        VBox box = new VBox();
        HBox h1 = new HBox();
        HBox h2 = new HBox();
        HBox h3 = new HBox();
        Label l1 = new Label("X: ");
        Label l2 = new Label("Y: ");
        Label l3 = new Label("Z: ");
        Label p1 = new Label();
        Label p2 = new Label();
        Label p3 = new Label();
        Slider x = new Slider(-Math.PI, Math.PI, 0);
        Slider y = new Slider(-Math.PI, Math.PI, 0);
        Slider z = new Slider(-Math.PI, Math.PI, 0);
        h1.getChildren().add(l1);
        h1.getChildren().add(x);
        h1.getChildren().add(p1);
        h2.getChildren().add(l2);
        h2.getChildren().add(y);
        h2.getChildren().add(p2);
        h3.getChildren().add(l3);
        h3.getChildren().add(z);
        h3.getChildren().add(p3);
        Button reset = new Button("Reset");
        reset.setOnAction(new EventHandler<ActionEvent>() {
            @Override
            public void handle(ActionEvent event) {
                enqueue(() -> player.setLocalRotation(new Quaternion().fromAngles(0f, 0f, 0f)));
                x.valueProperty().setValue(0f);
                y.valueProperty().setValue(0f);
                z.valueProperty().setValue(0f);
            }
        });
        ChangeListener listener = new ChangeListener<Number>() {
            @Override
            public void changed(ObservableValue<? extends Number> observableValue, Number oldValue, Number newValue) {
                enqueue(() ->
                        player.setLocalRotation(new Quaternion().fromAngles((float) x.getValue(), (float) y.getValue(), (float) z.getValue()))
                );
                float[] xyz = player.getLocalRotation().toAngles(new float[3]);
                p1.setText("X " + xyz[0] + " = " + x.getValue());
                p2.setText("Y " + xyz[1] + " = " + y.getValue());
                p3.setText("Z " + xyz[2] + " = " + z.getValue());
            }
        };
        x.valueProperty().addListener(listener);
        y.valueProperty().addListener(listener);
        z.valueProperty().addListener(listener);

        box.getChildren().add(h1);
        box.getChildren().add(h2);
        box.getChildren().add(h3);
        box.getChildren().add(reset);
        JavaFxUI.getInstance().attachChild(box);
        Node world = new Node("world");

        world.attachChild(player);
        rootNode.attachChild(world);

        cam.setLocation(new Vector3f(0, 10f, 50f));
        cam.setFrustumFar(5000f);
        flyCam.setMoveSpeed(20f);


        ActionListener tab = new ActionListener() {
            @Override
            public void onAction(String name, boolean isPressed, float tpf) {
                if (!isPressed) {
                    cursor = !cursor;
                    flyCam.setEnabled(cursor);
                }
            }
        };
        inputManager.addMapping("tab", new KeyTrigger(KeyInput.KEY_TAB));
        inputManager.addListener(tab, "tab");

    }

    public static void main(String[] args) {
        RotationTest main = new RotationTest();
        main.start();
    }
}
```

**Comment by stephengold** [comment]
Fixed in master at 9183f0e (v3.4.0-alpha3)
